### PR TITLE
Add aks support

### PR DESCRIPTION
**Pull request by boris-smidt-klarrio**
Adding support for tagging pvc's in AKS.
We have tested this code internally to make sure it works.

### CODE REVIEW

**Comment by mtougeron** [comment]
Thanks for the PR! I'll take a look at it over the next few days. While I do so, can you update the version of `github.com/Azure/azure-sdk-for-go/sdk/azidentity` to something newer? The version used has some vulnerabilities. Thanks!

**Comment by mtougeron** [comment]
so sorry, I got distracted and totally dropped the ball on reviewing this. I'll make it a priority for the next day or two.

**Comment by boris-smidt-klarrio** [comment]
@mtougeron We have this change running for quite some time now and it seems to function stable.
Can we get it merged?

**Comment by mtougeron** [comment]
Yes, I've been really distracted as of late. I'm really sorry for not getting this taken care of. It's a top priority for me to get this merged and released in the next day or two.

File: Dockerfile [hunk 1]
```diff
@@ -37,7 +37,8 @@ RUN addgroup -S k8s-pvc-tagger && adduser -S k8s-pvc-tagger -G k8s-pvc-tagger
 # Build a small image
 FROM scratch
 COPY --from=builder /etc/passwd /etc/passwd
-USER k8s-pvc-tagger
```
Comment (mtougeron) [inline Dockerfile:40]: Why did you remove the `USER` statement?

File: aks.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+	return subscription, resourceGroup, diskName, nil
+}
+
+func sanitizeLabelsForAzure(tags map[string]string) (DiskTags, error) {
+	diskTags := make(DiskTags)
+	if len(tags) > 50 {
+		return nil, ErrAzureTooManyTags
+	}
+	for k, v := range tags {
+		k = sanitizeKeyForAzure(k)
```
Comment (mtougeron) [inline aks.go:95]: After sanitizing the key it is possible that it could end up being a duplicate key. I'm on the fence as to whether it would be better to reject the tag if the key needs to be sanitized or if the tag created a duplicate. Which approach do you think is better?

File: aks.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	if len(fields) != 9 {
+		return "", "", "", errors.New("invalid volume id")
+	}
+	subscription = fields[2]
+	resourceGroup = fields[4]
+	diskName = fields[8]
+	return subscription, resourceGroup, diskName, nil
+}
+
+func sanitizeLabelsForAzure(tags map[string]string) (DiskTags, error) {
```
Comment (mtougeron) [inline aks.go:89]: Good call on sanitizing these values. I always forget this is necessary for Azure.

File: aks.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	return diskTags, nil
+}
+
+func sanitizeKeyForAzure(s string) string {
+	// remove forbidden characters
+	if strings.ContainsAny(s, `<>%&\?/`) {
+		for _, c := range `<>%&\?/` {
+			s = strings.ReplaceAll(s, string(c), "")
```
Comment (mtougeron) [inline aks.go:111]: If the invalid characters are being removed from the key it is possible to bypass this logic https://github.com/mtougeron/k8s-pvc-tagger/blob/main/kubernetes.go#L327-L339 by creating a tag name of `Kubernetes/Cluster: foo`

File: kubernetes.go [hunk 5]
```diff
@@ -252,6 +285,26 @@ func watchForPersistentVolumeClaims(ch chan struct{}, watchNamespace string) {
...
+
+	provisionedBy, ok := getProvisionedBy(annotations)
+	if !ok {
+		log.WithFields(log.Fields{"namespace": pvc.GetNamespace(), "pvc": pvc.GetName()}).Debugln("no volume.kubernetes.io/storage-provisioner annotation")
+		return false
+	}
+
+	switch provisionedBy {
+	case AZURE_DISK_CSI:
+		log.WithFields(log.Fields{"namespace": pvc.GetNamespace(), "pvc": pvc.GetName()}).Debugln(AWS_EBS_LEGACY + " volume")
```
Comment (mtougeron) [inline kubernetes.go:302]: Wrong constant `AWS_EBS_LEGACY` used here

File: aks.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
+package main
```
Comment (mtougeron) [inline aks.go:1]: nit: this and `aks_test.go` are the only files/variables that use `aks` instead of `azure` and this code could work for both `aks` and self-hosted k8s clusters on Azure. Could you rename this to be `azure.go` instead?

File: README.md [hunk 7]
```diff
@@ -117,6 +117,9 @@ gcloud iam roles create CustomDiskRole \
     --stage="GA"
 ```
 
+#### Azure rule
+The default role `Tag Contributor` can be used to configure the access rights for the pvc-tagger.
```
Comment (mtougeron) [inline README.md:121]: Can you add a bit about how this only support Azure `disk` csi volumes?

**Review by mtougeron — CHANGES_REQUESTED** [review]
This is great stuff. I really appreciate the addition. Again, I'm very sorry for how long this took me to review.

I have a few pieces of feedback, but the main thing of concern is the sanitization of the tags and how stripping the invalid characters can bypass other validation rules.

Otherwise, this is AWESOME!

**Comment by mtougeron** [comment]
@boris-smidt-klarrio btw, if these changes are too much of a pain for you, because of how long I took to get to a review, I'm okay with merging as-is and creating an RC release of the image and making the changes myself. Just let me know what you'd prefer.

**Comment by boris-smidt-klarrio** [comment]
@mtougeron I'll pull it to the end, there isn't any rush to this we are running an internal version of this image in the meantime.

File: aks.go (hunk 2 above)
Comment (boris-smidt-klarrio) [reply aks.go:95]: We use this in AWS and Azure so we don't really want to change the tags depending on the envrionment. 
But i will add an error in case there are duplicated keys, this should also tackle your other comment concerning `Kubernetes/Cluster: foo`

File: README.md (hunk 7 above)
Comment (boris-smidt-klarrio) [reply README.md]: I've documented the behaviors hopefully this description will stay up to date in the future.

**Comment by boris-smidt-klarrio** [comment]
One minor change I've made after reading the GCP note i changed the behavior to replace the invalid characters in azure with _  as wel.

File: aks.go (hunk 4 above)
Comment (boris-smidt-klarrio) [reply aks.go]: Its solved by rejecting duplicated tag keys during the sanitization.

**Review by mtougeron — APPROVED** [review]
Looks great! thank you!
